esp/platform/pluginloader.hpp

```diff
@@ -0,0 +1,184 @@
+    {
+        StringBuffer missing;
+        if (m_libraryXPath.isEmpty() && m_libraryDefault.isEmpty())
+        {
+            if (!missing.isEmpty())
+                missing.append(" and ");
+            missing.append("library name");
+        }
+        if (m_entryPointXPath.isEmpty() && m_entryPointDefault.isEmpty())
+        {
+            if (!missing.isEmpty())
+                missing.append(" and ");
+            missing.append("entry-point function name");
+        }
+        if (!missing.isEmpty())
+            ERRLOG("bad plugin loading configuration - missing %s", missing.str());
+    }
+
+    template <typename instance_t>
+    instance_t* create(const IPTree& configuration, Creator<instance_t> creator)
+    {
+        instance_t* instance = nullptr;
+        if (!creator)
+        {
+            ERRLOG("plugin creation failed - invalid creation hook function");
+            return nullptr;
+        }
+        else
+        {
+            try
+            {
+                entry_point_t entryPoint = lookup(configuration);
+                if (entryPoint)
+                    instance = creator(entryPoint);
+            }
+            catch (IException* e)
+            {
+                StringBuffer msg;
+                ERRLOG("plugin creation exception: %s", e->errorMessage(msg).str());
+            }
+            catch (...)
+            {
+                ERRLOG("plugin creation exception");
+            }
+        }
+        return instance;
+    }
+
+    entry_point_t lookup(const IPTree& configuration)
+    {
+        const char* libraryName = nullptr;
+        const char* entryPointName = nullptr;
+        if (!m_libraryXPath.isEmpty())
+            libraryName = configuration.queryProp(m_libraryXPath);
+        if (!m_entryPointXPath.isEmpty())
+            entryPointName = configuration.queryProp(m_entryPointXPath);
+        return lookup(libraryName, entryPointName);
+    }
+
+    entry_point_t lookup(const char* libraryName = nullptr, const char* entryPointName = nullptr)
+    {
+        bool canFind = true;
+        if (isEmptyString(libraryName))
+        {
+            if (m_libraryDefault.isEmpty())
+            {
+                ERRLOG("plugin loader name lookup failed - no library name and no default value");
+                canFind = false;
+            }
+            else
+                libraryName = m_libraryDefault;
+        }
+        if (isEmptyString(entryPointName))
+        {
+            if (m_entryPointDefault.isEmpty())
+            {
+                ERRLOG("plugin loader name lookup failed - no entry point name and no default value");
+                canFind = false;
+            }
+            else
+                entryPointName = m_entryPointDefault;
+        }
+        return (canFind ? find(libraryName, entryPointName) : nullptr);
+    }
+
+private:
+    using EntryPointMap = std::map<std::string, entry_point_t>;
+    struct Plugin
+    {
+        HINSTANCE library = nullptr;
```

Comment:
Is the shared object freed at some point? Maybe better to use the "SharedObject" jlib class? Need to be careful not to free the library when its procedures are or will still be used.

Reply:
Short answer is no. This is essentially a refactoring of existing code to be more generic, and the shared object was not being freed.

The shared object cannot be freed until all instances created by it are released. The public visibility of the entry point functions prevents this class from tracking all instances. Even if it could track instances, coupling the external destruction of a tracked instance to an embedded SharedObject instance could get very messy.

If the now static instance became a member of the service instance, each service instance could clean up after itself. The success of this approach could be dependent upon the order of deletion, which might not be entirely in the service instance's control.

How important is this existing issue and, if important enough to address now, do you have recommendations for safely tracking this relationship?

Reply:
I've created HPCC-25116 to track this concern.